src/Rules/Arrays/InvalidKeyInArrayDimFetchRule.php

```diff
@@ -54,15 +56,34 @@ public function processNode(Node $node, Scope $scope): array
 		}
 
 		$isSuperType = AllowedArrayKeysTypes::getType()->isSuperTypeOf($dimensionType);
-		if ($isSuperType->yes() || ($isSuperType->maybe() && !$this->reportMaybes)) {
-			return [];
+		$isOk = $isSuperType->yes() || ($isSuperType->maybe() && !$this->reportMaybes);
+		if (!$isOk) {
+			return [
+				RuleErrorBuilder::message(
+					sprintf('%s array key type %s.', $isSuperType->no() ? 'Invalid' : 'Possibly invalid', $dimensionType->describe(VerbosityLevel::typeOnly())),
+				)->identifier('offsetAccess.invalidOffset')->build(),
+			];
+		}
+
+		if ($this->phpVersion->getVersionId() >= 80100) {
+			$isFloat = $dimensionType->isFloat();
+
+			if ($isFloat->yes()) {
```

Comment:
We need to use RuleLevelHelper::findTypeToCheck here, it's going to help us on level 9 and also with benevolent union types.

Case in point: `$mixed - 1` is already benevolent union type so this problem would already be solved: https://phpstan.org/r/5af2741f-b958-42f7-9248-435416d1b584